### PR TITLE
feature: add support for additional cache keys option

### PR DESCRIPTION
**Pull request by Rohit-L**
PR for issue #303 

This adds an option for user to pass additional keys to the default cache key generator for more customization, allowing the user to have more control, if needed, over what hits/misses in the cache.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -112,6 +113,35 @@ Default path to cache directory: `node_modules/.cache/uglifyjs-webpack-plugin`.
...
+
+#### `{Object}`
+
+**webpack.config.js**
+```js
+[
+  new UglifyJsPlugin({
+    cache: true,
+    additionalCacheKeys: {
+      env: 'development',
```
Comment (evilebottnawi) [inline README.md]: Bad example, many people can use this from `README`, just use `myCustomVariable` name or something like

File: test/cache-options.test.js [hunk 2]
```diff
@@ -436,4 +436,152 @@ describe('when options.cache', () => {
...
+
+    beforeEach(() => {
+      const pluginEnvironment = new PluginEnvironment();
+      const compilerEnv = pluginEnvironment.getEnvironmentStub();
+      compilerEnv.context = '';
+
+      const plugin = new UglifyJsPlugin({
+        cache: true,
+        additionalCacheKeys: {
+          env: 'development',
```
Comment (evilebottnawi) [inline test/cache-options.test.js]: Please also use other variable

**Comment by Rohit-L** [comment]
👍 Updated with suggestions from @evilebottnawi

**Comment by evilebottnawi** [comment]
@ooflorent i am not fan also, but it should be documented because in some case it is only one variant to invalidate cache

**Comment by shellscape** [comment]
In the past I've taken large, complex portions of a README and moved them out into separate `.md` files in a `./docs` directory. That might be considered to keep the README a little smaller. I would like to review this PR but I know practically nothing about how the plugin works, so I'm afraid that my review wouldn't be very useful.

**Comment by evilebottnawi** [comment]
@shellscape i think we can do this in other PR, not here

**Comment by Rohit-L** [comment]
@evilebottnawi @ooflorent what do you think about making `additionalCacheKeys` a function that returns the object of additional keys instead?

I just realized, for my use-case, I need access to `file` when defining my additional cache keys and this implementation wouldn't allow that.

I'm proposing something similar to how [cache-loader](https://github.com/webpack-contrib/cache-loader) defines their `cacheKey` option.

**Comment by evilebottnawi** [comment]
@Rohit-L can you provide example?

**Comment by evilebottnawi** [comment]
Close in favor https://github.com/webpack-contrib/uglifyjs-webpack-plugin/pull/320

**Pull request closed without merging**